text/0000-pi-types.md

```diff
@@ -0,0 +1,644 @@
+           | O0..On+r                      // Object type
+           | [T]                           // Slice type
+           | for<r..> fn(T1..Tn) -> T0     // Function pointer
+           | <P0 as Trait<P1..Pn>>::Id     // Projection
+    +      | C                             // const types
+    +    F = c                             // const fn name
+    +    C = E                             // Pi constructed const type
+         P = r                             // Region name
+           | T                             // Type
+         O = for<r..> TraitId<P1..Pn>      // Object type fragment
+         r = 'x                            // Region name
+    +    E = F(E)                          // Constant function application.
+    +      | p                             // const type parameter
+    +      | [...]                         // etc.
+
+Note that the `const` syntax is only used when declaring the parameter.
+
+## `impl` unification
+
+Only one `where` bound can be specified on each disjoint implementations (for
+possible extensions, see below).
+
+To find the right implementation, we use the data from the type inference (see
+the inference rules above). Since the parameters are, in fact, not much
+semantically different from normal generic parameters, we can resolve it in a
+normal manner (that is, by treating the value parameters as if they were actual
+type parameters).
+
+Likewise are disjointness checks based on structural equality. That is, we only
+care about structural equality, not `Eq` or something else. This allows us to
+reason more rigorously about the behavior.
+
+Since not all parameters' edges are necessarily the identity function,
+dispatching these would be undecidable. A way to solve this problem is to
+introduce some syntax allowing to specify the `impl` parameters. This is not
+something we consider in this proposal, but a secondary RFC can introduce these.
+
+## Division by zero
+
+If some function contain a constexpr divisor, dependent on some value parameter
+of the function, that is (`a / f(x)`), the compiler must ensure that the bound
+implies that `f(x) != 0`.
+
+## An example
+
+This is the proposed syntax:
+
+```rust
+use std::{mem, ptr};
+
+// We start by declaring a struct which is value dependent.
+struct Array<const n: usize, T> {
+    // `n` is a constexpr, sharing similar behavior with `const`s, thus this
+    // is possible.
+    content: [T; n],
+}
+
+// We are interested in exploring the `where` clauses and Π-constructors:
+impl<const n: usize, T> Array<n, T> {
+    // This is simple statically checked indexing.
+    fn checked_index<const i: usize>(&self) -> &T where i < n {
+        //                 note that this is constexpr  ^^^^^
+        unsafe { self.content.unchecked_index(i) }
+    }
+
+    // "Push" a new element, incrementing its length **statically**.
+    fn push(self, elem: T) -> Array<n + 1, T> {
```

Comment:
You can just call any `const fn` there that does whatever you want:

    const fn foo() -> usize { 
        let n = "foo"; 
        let m = foo.len(); 
        if m > 2 { m * 3 } else { 42 } 
    } 
    Array<foo(), T>

Or maybe even use a closure:

    Array<const || {
            let n = "foo"; 
            let m = foo.len(); 
            if m > 2 { m * 3 } else { 42 } 
    }(), T>

@ticki I don't think that the RFC mentions `const` closures yet, it probably should.